### ISSUE TITLE
MultiTouch behaving incorrectly with MOE backend

### ISSUE THREAD

**Issue by majiddaeinejad** [libgdx/libgdx#4114]
#### Version of LibGDX

1.9.3
#### Version of Multi-OS Engine

1.0.648
#### Affected platforms

iOS (MOE)
#### Tested Devices

iPhone 5s and iPhone 6s

The issue is about touching with two fingers or more. The InputListener doesn't work correctly in iOS version after migration to Multi-OS engine. The 'touchup' for one of fingers has not been called in iOS version if display touched with two fingers or more together simultaneously, only 'touchdown' occured.
Simply you can test it by this code:

```
Gdx.input.setInputProcessor(new InputProcessor() {
            @Override
            public boolean keyDown(int keycode) {
                return false;
            }

            @Override
            public boolean keyUp(int keycode) {
                return false;
            }

            @Override
            public boolean keyTyped(char character) {
                return false;
            }

            @Override
            public boolean touchDown(int screenX, int screenY, int pointer, int button) {
                Gdx.app.log("touchdown", "touchdown");
                return true;
            }

            @Override
            public boolean touchUp(int screenX, int screenY, int pointer, int button) {
                Gdx.app.log("touchup", "touchup");
                return true;
            }

            @Override
            public boolean touchDragged(int screenX, int screenY, int pointer) {
                return false;
            }

            @Override
            public boolean mouseMoved(int screenX, int screenY) {
                return false;
            }

            @Override
            public boolean scrolled(int amount) {
                return false;
            }
        });

```

**Comment by Tom-Ski** [comment]
Please fill in the Issue template you are shown before submitting an issue.  What version of LibGDX are you using? What version of MOE are you using? What device are you testing on?

**Comment by majiddaeinejad** [comment]
@Tom-Ski I have edited this issue according to the issue template.

**Comment by Tom-Ski** [comment]
Thanks, Ill have a look soon.  Looks like its an issue with touch up events, or maybe multi touch interfering somehow.

**Comment by vbousquet** [comment]
I can confirm this bug since I have the same. I've ported my game to both RoboVM (MobiDevelop branch) and MOE and the bug only appears for the MOE branch. It happens for single touch as well as for multi touch.

**Comment by vbousquet** [comment]
After investigation, there seems to be a bug in the MOE version of IOSInput which drops event.
The following extract shows that it use ```break``` instead of ```continue``` dropping the following events if any. It will happen for fast click (my case) or for multitouch (original poster case).

```

	void processEvents () {
		synchronized (touchEvents) {
			justTouched = false;
			for (TouchEvent event : touchEvents) {
				currentEvent = event;
				if (event.phase == UITouchPhase.Began) {
					if (inputProcessor != null) inputProcessor.touchDown(event.x, event.y, event.pointer, Buttons.LEFT);
					if (numTouched == 1) justTouched = true;
					break;
				}
				if (event.phase == UITouchPhase.Cancelled || event.phase == UITouchPhase.Ended) {
					if (inputProcessor != null) inputProcessor.touchUp(event.x, event.y, event.pointer, Buttons.LEFT);
					break;
				}
				if (event.phase == UITouchPhase.Moved || event.phase == UITouchPhase.Stationary) {
					if (inputProcessor != null) inputProcessor.touchDragged(event.x, event.y, event.pointer);
					break;
				}
			}
			touchEventPool.freeAll(touchEvents);
			touchEvents.clear();
		}
	}
```

The easy fix is either to go for a continue, or just to copy the code for the same method in the RoboVM backend ;

```
	void processEvents () {
		synchronized (touchEvents) {
			justTouched = false;
			for (TouchEvent event : touchEvents) {
				currentEvent = event;
				switch (event.phase) {
				case Began:
					if (inputProcessor != null) inputProcessor.touchDown(event.x, event.y, event.pointer, Buttons.LEFT);
					if (numTouched == 1) justTouched = true;
					break;
				case Cancelled:
				case Ended:
					if (inputProcessor != null) inputProcessor.touchUp(event.x, event.y, event.pointer, Buttons.LEFT);
					break;
				case Moved:
				case Stationary:
					if (inputProcessor != null) inputProcessor.touchDragged(event.x, event.y, event.pointer);
					break;
				}
			}
			touchEventPool.freeAll(touchEvents);
			touchEvents.clear();
		}
	}
```

I do not have a setup (I develop for iOs only in AndroidStudio and modify LibGdx only under Eclipse) to validate this, so this is not a PR for the moment but just an unchecked fix proposal.

**Comment by vbousquet** [comment]
I have just proposed the PR since this fix the issue. @majiddaeinejad can you test and confirm that this fix it for you too ?

**Comment by majiddaeinejad** [comment]
Hey @vbousquet, I was using gdxsetup for creating my project, anyway I'm going to test in with your solution.

**Comment by psiegel** [comment]
I just hit this bug myself and came here to report it, but I see you guys are already on it.  FWIW, I am able to produce this bug in my project without even using multi-touch.  A simple quick tap that should produce a touchDown followed by a touchUp event can drop the touchUp due to this bug.  I see things like buttons sticking down and never firing their onClick events as a result.  Looking forward to seeing that PR merged in.

**Comment by vbousquet** [comment]
There are a lot of PR at the moment to review so the core team may need some time to merge it.
If you need it, you can just work from the source, it is quite easy (I did it to validate the PR) ;
1. clone the repo and copy this fix (2 lines of code)
2. import the build.gradle from the cloned repo in Android Studio as a new project
3. add ```apply plugin: "maven"``` to the part which goes for all project configuration
4. if you encounter error with the robovm part (which is not the MobiDevelop one), just comment out the robovm project (in settings.gradle and build.gradle)
5. launch a ```gradle install```
6. you now have the latest branch with the fix installed in your local maven repo available as 1.9.4-SNAPSHOT branch, enjoy !

**Comment by psiegel** [comment]
Thanks vousquet, I have pushed forward with exactly that.  FWIW, I was having trouble getting my project to use the locally build libgdx but didn't want to clutter up this issue report with further questions, so I posted about it on the forums here:
http://www.badlogicgames.com/forum/viewtopic.php?f=11&t=22972
And in the process of doing so, solved my own problem.  So, cross-posting the link here for posterity.  Nothing else to see here, looking forward to seeing this fix make its way into the official 1.9.4.
